File: python/src/toolrate_mcp/tools/categories.py
"""``toolrate_categories`` — list canonical categories with counts."""
from __future__ import annotations

import json
from typing import Any

from ..client import api_request
from ..errors import raise_tool_error

NAME = "toolrate_categories"

DESCRIPTION = "Get the canonical category names + tool counts. Useful to filter toolrate_search_tools or toolrate_hidden_gems."


async def handler() -> str:
    try:
        body, rate_limit = await api_request(method="GET", path="/v1/tools/categories")
    except Exception as e:
        raise_tool_error(e)

    result: dict[str, Any] = {**body}
    if rate_limit is not None:
        result["_meta"] = {"rate_limit": rate_limit}
    return json.dumps(result, indent=2)

File: typescript/src/tools/categories.ts
import { apiRequest } from "../client.js";
import { mcpError } from "../errors.js";
import type { RawCategoriesResponse } from "../schema.js";
import { categoriesInputSchema } from "../types.js";

export const name = "toolrate_categories";

export const config = {
  title: "List all tool categories with counts",
  description: "Get the canonical category names + tool counts. Useful to filter toolrate_search_tools or toolrate_hidden_gems.",
  inputSchema: categoriesInputSchema,
};

export async function handler(_input: Record<string, never>) {
  try {
    const { body, rateLimit } = await apiRequest<RawCategoriesResponse>({
      method: "GET",
      path: "/v1/tools/categories",
    });

    const result = {
      ...body,
      ...(rateLimit ? { _meta: { rate_limit: rateLimit } } : {}),
    };

    return {
      content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
    };
  } catch (err) {
    return mcpError(err);
  }
}
